src/frameworks/compose/discover/runtime/node.ts

```diff
@@ -0,0 +1,216 @@
+import { readOrNull } from "../filesystem";
+import { FileSystem, FrameworkSpec, Runtime } from "../types";
+import { RuntimeSpec } from "../types";
+import { frameworkMatcher } from "../frameworkMatcher";
+import { LifecycleCommands } from "../types";
+import { Command } from "../types";
+import { FirebaseError } from "../../../../error";
+import { logger } from "../../../../../src/logger";
+
+export interface PackageJSON {
+  dependencies?: Record<string, string>;
+  devDependencies?: Record<string, string>;
+  scripts?: Record<string, string>;
+  engines?: Record<string, string>;
+}
+type PackageManager = "npm" | "yarn";
+
+const supportedNodeVersion = 18;
+const NODE_RUNTIME_ID = "nodejs";
+const PACKAGE_JSON = "package.json";
+const YARN_LOCK = "yarn.lock";
+const PACKAGE_LOCK_JSON = "package-lock.json";
+
+export class NodejsRuntime implements Runtime {
+  private readonly runtimeRequiredFiles: string[] = [PACKAGE_JSON];
+  private readonly contentCache: Record<string, boolean> = {};
+
+  // Checks if the codebase is using Node as runtime.
+  async match(fs: FileSystem): Promise<boolean | null> {
+    const areAllFilesPresent = await Promise.all(
+      this.runtimeRequiredFiles.map((file) => fs.exists(file))
+    );
+
+    return areAllFilesPresent.every((present) => present);
+  }
+
+  getRuntimeName(): string {
+    return NODE_RUNTIME_ID;
+  }
+
+  getNodeImage(engine: Record<string, string> | undefined): string {
+    // If no version is mentioned explicitly, assuming application is compatible with latest version.
+    if (!engine) {
```

Comment:
you'd want to also check that `engine.node` isn't undefined.

Reply:
Done